File: js/shared/date_parser.js
/**
 * Robust multi-format date parser (browser version).
 * Extracts {year, month} from Date objects, Excel serial numbers,
 * ISO dates, Chinese dates, and year-month formats.
 */

function parseDate(val) {
  if (val == null || val === '') return { year: null, month: null };

  // Date object
  if (val instanceof Date) {
    return { year: val.getFullYear(), month: val.getMonth() + 1 };
  }

  // Number (Excel serial)
  if (typeof val === 'number') {
    if (val > 3000 && val < 100000) {
      var base = new Date(1899, 11, 30);
      base.setDate(base.getDate() + Math.floor(val));
      return { year: base.getFullYear(), month: base.getMonth() + 1 };
    }
    if (val >= 2000 && val <= 2100) return { year: Math.floor(val), month: null };
    return { year: null, month: null };
  }

  var s = String(val).trim();
  if (!s) return { year: null, month: null };

  // Try various date formats
  var m;
  m = s.match(/^(\d{4})-(\d{1,2})-(\d{1,2})$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})\/(\d{1,2})\/(\d{1,2})$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})\.(\d{1,2})\.(\d{1,2})$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})-(\d{1,2})$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})\/(\d{1,2})$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})年(\d{1,2})月(\d{1,2})日?$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }
  m = s.match(/^(\d{4})年(\d{1,2})月$/);
  if (m) { var y=parseInt(m[1]), mo=parseInt(m[2]); if(y>=2000&&y<=2100&&mo>=1&&mo<=12) return {year:y, month:mo}; }

  return { year: null, month: null };
}

File: src/shared/date_parser.py
"""
Robust multi-format date parser.
Extracts (year, month) from datetime objects, Excel serial numbers,
ISO dates, Chinese dates, and year-month formats.
"""
import re
import datetime as _dt
import pandas as pd
import numpy as np


def parse_date(val):
    """
    Parse a value into (year, month).
    Supports: '2024-01-15', '2024/1/15', '2024年1月', '2024-01',
              Excel serial numbers (45292), datetime objects, Timestamp.
    Returns (year, month) or (None, None).
    """
    if val is None:
        return None, None

    # pandas Timestamp / Python datetime
    if isinstance(val, (_dt.datetime, _dt.date, pd.Timestamp)):
        return val.year, val.month

    # Numeric (Excel serial: 1900-01-01 = 1)
    if isinstance(val, (int, float, np.integer, np.floating)):
        if 3000 < val < 100000:  # Reasonable Excel date range (1900-2170)
            try:
                base = _dt.datetime(1899, 12, 30)
                dt = base + _dt.timedelta(days=int(val))
                return dt.year, dt.month
            except Exception:
                pass
        if 2000 <= val <= 2100:  # Might be just a year
            return int(val), None
        return None, None

    s = str(val).strip()
    if not s:
        return None, None

    # Try various formats
    formats = [
        # ISO variants
        (r'^(\d{4})-(\d{1,2})-(\d{1,2})$', 1, 2),
        (r'^(\d{4})/(\d{1,2})/(\d{1,2})$', 1, 2),
        (r'^(\d{4})\.(\d{1,2})\.(\d{1,2})$', 1, 2),
        # Year-Month only
        (r'^(\d{4})-(\d{1,2})$', 1, 2),
        (r'^(\d{4})/(\d{1,2})$', 1, 2),
        # Chinese
        (r'^(\d{4})年(\d{1,2})月(\d{1,2})日?$', 1, 2),
        (r'^(\d{4})年(\d{1,2})月$', 1, 2),
    ]

    for pattern, yi, mi in formats:
        m = re.match(pattern, s)
        if m:
            y, mo = int(m.group(yi)), int(m.group(mi))
            if 2000 <= y <= 2100 and 1 <= mo <= 12:
                return y, mo

    return None, None
